### ISSUE TITLE
Disks mounted after Docker service start are inaccessible by the containers

### ISSUE THREAD

**Issue by gfjardim** [limetech/dynamix-6.2#4]
This is a well know behaviour of Docker. Because the daemon is started using unshare, the mounts namespace is read-only to the child containers and disks mounted after the daemon start are invisible to them.

My proposal is to change `unshare` from unshare mounts namespace `-m` to `--propagation slave`. Then, we have to share the `/mnt` directory before emhttp starts by adding this command - `/sbin/mount --bind /mnt /mnt && /sbin/mount --make-rshared /mnt` -  to `/etc/rc.d/rc.local` file.

After that, all disks mounted under /mnt/ will be `shared` by default, so we don't need to share individual directories.

The proposed `rc.docker` file:

```
#!/bin/sh

# Short-Description:  Create lightweight, portable, self-sufficient containers.
# Description:
#  Docker is an open-source project to easily create lightweight, portable,
#  self-sufficient containers from any application. The same container that a
#  developer builds and tests on a laptop can run at scale, in production, on
#  VMs, bare metal, OpenStack clusters, public clouds and more.

# modified for unRAID

PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin

BASE=docker

UNSHARE=/usr/bin/unshare
DOCKER=/usr/bin/$BASE
DOCKER_PIDFILE=/var/run/$BASE.pid
DOCKER_LOG=/var/log/docker.log

## Set defaults used by the docker daemon
## These are flags passed after `docker daemon`
DOCKER_OPTS="--storage-driver=btrfs"

# Docker settings for unRAID
DOCKER_IMAGE_FILE=
DOCKER_IMAGE_SIZE=10
DOCKER_ENABLED="no"

if [ -f /boot/config/docker.cfg ]; then
  . /boot/config/docker.cfg
else
  echo "# Generated settings:" >> /boot/config/docker.cfg
  echo "DOCKER_OPTS=\"$DOCKER_OPTS\"" >> /boot/config/docker.cfg
  echo "DOCKER_IMAGE_FILE=\"$DOCKER_IMAGE_FILE\"" >> /boot/config/docker.cfg
  echo "DOCKER_IMAGE_SIZE=\"$DOCKER_IMAGE_SIZE\"" >> /boot/config/docker.cfg
  echo "DOCKER_ENABLED=\"$DOCKER_ENABLED\"" >> /boot/config/docker.cfg
fi

# Less verbose logging by default
DOCKER_OPTS="--log-level=error $DOCKER_OPTS"

export DOCKER_RAMDISK=true

# Check docker is present
if [ ! -x $DOCKER ]; then
	echo "$DOCKER not present or not executable"
	exit 1
fi

# Verify if docker daemon running
is_docker_running(){
  [ ! -S /var/run/docker.sock ] && return 1
  if [[ $(docker info 2>&1) =~ "Cannot connect to the Docker daemon" ]]; then return 1; else return 0; fi
}

# Wait 10s to daemon start
wait_daemon(){
  for n in $(seq 10); do
    if is_docker_running; then return 0; else sleep 1; fi
  done
  return 1
}

# Containers Management
# All existing containers
all_containers(){
  while IFS=$'\n' read -ra LINES; do
    for LINE in "${LINES[@]}"; do
      IFS="  " read -ra VALUES <<< "$LINES"
      echo -e "${VALUES[@]:(-1)}\n"
    done
  done <<< "$(docker ps -a 2>/dev/null | grep -v CONTAINER)"
}

# Running containers
running_containers(){
  while IFS=$'\n' read -ra LINES; do
    for LINE in "${LINES[@]}"; do
      IFS="  " read -ra VALUES <<< "$LINES"
      echo -e "${VALUES[@]:(-1)}\n"
    done
  done <<< "$(docker ps 2>/dev/null | grep -v CONTAINER)"
}

# Is container running?
is_container_running(){
  local CONTAINER
  for CONTAINER in $(running_containers); do
    if [[ $CONTAINER == $1 ]]; then
      return 0
    fi
  done
  return 1
}

# Does the container exist?
container_exist(){
  local CONTAINER
  for CONTAINER in $(all_containers); do
    if [[ $CONTAINER == $1 ]]; then
      return 0
    fi
  done
  return 1
}

do_container_paths() {
  local container=$1
  # Deal with host directories
  for hostpath in $(docker inspect --format "{{ range .Mounts }} \"{{ .Source }}\" {{ end }}" "$container"); do
    hostpath=$(echo $hostpath|tr -d '"')
    # Make sure hostpath exist
    if [ ! -e "$hostpath" ]; then
      mkdir -p "$hostpath"
    fi
  done
}

# Start containers
start_containers(){
  local CONTAINER
  CMD="docker start"
  wait_daemon
  if ! is_docker_running; then return 1; fi
  if [ -f /var/lib/docker/unraid-autostart ]; then
    while read CONTAINER; do
      if container_exist $CONTAINER && ! is_container_running $CONTAINER; then 
        do_container_paths $CONTAINER
        OUT=$(eval "$CMD $CONTAINER" 2>&1)
        if  [[ $OUT =~ "Error:" ]]; then
          echo -n $CONTAINER": "$OUT
        else
          echo "$CONTAINER: started succesfully!"
        fi
      fi
    done </var/lib/docker/unraid-autostart
  fi
}

docker_start() {
  if is_docker_running; then
    echo "$DOCKER is already running"
    return 1
  fi
  if [ "$DOCKER_ENABLED" != "yes" ]; then
    echo "$DOCKER is not enabled"
    exit 1
  fi
  if ! mountpoint /var/lib/docker &> /dev/null ; then
    echo "no image mounted at /var/lib/docker"
    exit 1
  fi
  echo "starting $BASE ..."
  if [ -x ${DOCKER} ]; then
    # If there is an old PID file (no docker running), clean it up:
    if [ -r ${DOCKER_PIDFILE} ]; then
      if ! ps axc | grep docker 1> /dev/null 2> /dev/null ; then
        echo "Cleaning up old ${DOCKER_PIDFILE}."
        rm -f ${DOCKER_PIDFILE}
      fi
    fi
    nohup "${UNSHARE}" --propagation slave -- ${DOCKER} daemon -p ${DOCKER_PIDFILE} ${DOCKER_OPTS} >> ${DOCKER_LOG} 2>&1 &
  fi
}

# Stop docker:
docker_stop() {
  echo "stopping $BASE ..."
  # If there is no PID file, ignore this request...
  if [ -r ${DOCKER_PIDFILE} ]; then
    RUNNING=$(docker ps -q)
    if [[ $RUNNING ]]; then
      docker stop $RUNNING
    fi
    kill $(cat ${DOCKER_PIDFILE})
    # must ensure daemon has stopped before unmounting image volume
    sleep 1
    for i in {1..14}; do
      if [ -r ${DOCKER_PIDFILE} ]; then
        echo "waiting for docker to die..."
        sleep 1
      fi
    done
    if [ -r ${DOCKER_PIDFILE} ]; then
      echo "docker will not die!"
      exit 1
    fi
  fi
  # tear down the bridge
  ifconfig docker0 down 2> /dev/null
  brctl delbr docker0 2> /dev/null
}

# Restart docker:
docker_restart() {
	docker_stop
	sleep 1
	docker_start
	start_containers
}

case "$1" in
'start')
  docker_start
  start_containers
  ;;
'stop')
  docker_stop
  ;;
'restart')
  docker_restart
  ;;
'status')
  if [ -f ${DOCKER_PIDFILE} ] && ps -o cmd $(cat ${DOCKER_PIDFILE}) | grep -q $BASE ; then
	  echo "status of $BASE: running"
	  echo "running containers:" $(running_containers)
  else
	  echo "status of $BASE: stopped"
	  exit 1
  fi
  ;;
*)
  echo "usage $0 start|stop|restart|status"
esac

exit 0
```

Please note that I already added a **do_container_paths** function to deal with the deprecation of directory creation by the daemon.

The option to map a shared volume is already pushed in a PR.

Thanks for your consideration.

**Comment by gfjardim** [comment]
Just some testings:

1) With normal RW mode:

![With normal RW mode](https://i.imgur.com/IdhK2Cc.png)

2) With RW/Shared mode:

![With normal RW mode](https://i.imgur.com/PFr2BXN.png)

In both testings, I mounted the INTEL_SSDSC2CT120A3_CVMP21350057120BGN after the container started. 

Note that in the first image, the directory was created but the mount target was not propagated because the volume mounting (/mnt) was not shared. The correct behavior should be that shown on the second picture.

**Comment by limetech** [comment]
Interesting... A couple questions:

Couldn't the bind mount be specified simply as:

`/sbin/mount --bind --make-rshared /mnt /mnt`

And just to verify, if there are no other mount points under /mnt at the time the above command is executed (during startup), later mount points created will "inherit" the "--make-shared" flag?

**Comment by gfjardim** [comment]
That command did the trick (tested on the GO file).

Just after boot:

![Image](https://i.imgur.com/LiBOlX9.png)

After the array starts:

![image](https://i.imgur.com/gisCL4a.png)

After mounting some disks using Unassigned Devices:

![image](https://i.imgur.com/pUD0yC2.png)


The pertinent discussion topic about this is [here](https://github.com/docker/docker/pull/17034).

**Comment by gfjardim** [comment]
Found this awesome documentation about this: 

https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt

**Comment by eschultz** [comment]
Closing, this was implemented in 6.2.0-Beta19.  Thanks @gfjardim
